File: pset6/credit/credit.py
from cs50 import get_int

cardNumber = get_int("Number: ")
startNumber = 0
digit = 0
digitCount = 0
luhn = 0

while digitCount < 16 and cardNumber != 0:
    digit = int(cardNumber % 10)

    if (int)(digitCount % 2) == 0:
        luhn += digit
    else:
        if digit * 2 >= 10:
            luhn += (int)(digit * 2 % 10) + (int)(digit * 2 / 10)
        else:
            luhn += (int)(digit * 2)

    cardNumber = int(cardNumber / 10)

    if cardNumber < 100 and cardNumber >= 10:
        startNumber = cardNumber

    digitCount += 1

luhn = int(luhn % 10)

if digitCount == 15 and luhn == 0 and (startNumber == 34 or startNumber == 37):
    print("AMEX")
elif digitCount == 16 and luhn == 0 and (startNumber >= 51 and startNumber <= 55):
    print("MASTERCARD")
elif (digitCount == 13 or digitCount == 16) and luhn == 0 and (startNumber >= 40 and startNumber < 50):
    print("VISA")
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// prompts the user for a credit card number and then reports (via printf) whether it is a valid American Express, MasterCard, or Visa card number
int main(void)
{
    long cardNumber = get_long("Number: ");
    int startNumber = 0;
    int digit = 0;  // current digit
    int digitCount = 0;
    int luhn = 0;

    // go through digits
    for (digitCount = 0; digitCount < 16 && cardNumber != 0; digitCount++)
    {
        // get last digit
        digit = cardNumber % 10;
        if (digitCount % 2 == 0)
        {
            luhn += digit;
        }
        else
        {
            if (digit * 2 >= 10)
            {
                luhn += (digit * 2 % 10) + (digit * 2 / 10);
            }
            else
            {
                luhn += digit * 2;
            }
        }

        // remove last digit
        cardNumber /= 10;

        // get starting number
        if (cardNumber < 100 && cardNumber > 10)
        {
            startNumber = cardNumber;
        }
    }

    // luhn == 0 if valid
    luhn = luhn % 10;

    // print results
    if (digitCount == 15 && luhn == 0 && (startNumber == 34 || startNumber == 37))
    {
        printf("AMEX\n");
    }
    else if (digitCount == 16 && luhn == 0 && (startNumber >= 51 && startNumber <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((digitCount == 13 || digitCount == 16) && luhn == 0 && (startNumber >= 40 && startNumber < 50))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}